### ISSUE TITLE
Finetuning on basins

### ISSUE THREAD

**Issue by tommylees112** [neuralhydrology/neuralhydrology#18]
Hi guys!  (repeating an email on here - realised that the formatting and links are really helpful, ignore the email!)

I wanted to finetune some models on specific basins. 

# What arguments do i need in the configs to finetune?

# 1) `base_run_dir`
I found this information in the docs:
```
neuralhydrology.nh_run.finetune(config_file: pathlib.Path = None, gpu: int = None)
Finetune a pre-trained model.

Parameters
config_file (Path, optional) – Path to an additional config file. Each config argument in this file will overwrite the original run config. The config file for finetuning must contain the argument base_run_dir, pointing to the folder of the pre-trained model.

gpu (int, optional) – GPU id to use. Will override config argument ‘device’.
```
So I need to specify a new argument: `base_run_dir`

# 2) `finetune_modules`
[basetrainer.py L106-113](https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/training/basetrainer.py#L106-L113)
```python
        if isinstance(self.cfg.finetune_modules, list):
            for module_part in self.cfg.finetune_modules:
                if module_part in self.model.module_parts:
                    module = getattr(self.model, module_part)
                    for param in module.parameters():
                        param.requires_grad = True
                else:
                    unresolved_modules.append(module_part)

```
I guess therefore I need to assign values to the argument `finetune_modules` in the yaml file. What parameters can this argument take? Is it "head" or "lstm" ?

```yml
finetune_modules: 
```

# 3) `train_basin_file`
```yaml
train_basin_file: /home/tommy/neuralhydrology/data/finetune_basin_list.txt
validation_basin_file: /home/tommy/neuralhydrology/data/finetune_basin_list.txt
test_basin_file: /home/tommy/neuralhydrology/data/finetune_basin_list.txt
```

Are there any other arguments that I have overlooked or got wrong?

Thanks guys! 
Tommy

**Comment by gauchm** [comment]
Hi Tommy,
yes, you need to specify a `base_run_dir` (the run you want to finetune) and `finetune_modules`.
By default, finetuning will start with the last saved model from `base_run_dir`. You can change this behavior by specifying a model path in `checkpoint_path`. E.g., if you think your base model was already overfitting, you can use an earlier checkpoint this way.

**`finetune_modules`** expects a list of the parts of the model you want to finetune.
Each model in the codebase defines which modules it consists of in a variable called "module_parts". E.g., the normal cudalstm consists of ["lstm" and "head"](https://github.com/neuralhydrology/neuralhydrology/blob/49732110d500300589aac2284427bd51a20a1b68/neuralhydrology/modelzoo/cudalstm.py#L32), and the EA-LSTM has ["input_gate", "dynamic_gates", and "head"](https://github.com/neuralhydrology/neuralhydrology/blob/49732110d500300589aac2284427bd51a20a1b68/neuralhydrology/modelzoo/ealstm.py#L36).

So, if you've trained a normal LSTM ("cudalstm") with a regression head, and you want to finetune the whole thing, you'd pass
```yaml
finetune_modules:
  - lstm
  - head
```

Whereas, if you only want to finetune the head, you'd do
```yaml
finetune_modules:
  - head
```
(or, more concisely, `finetune_modules: head`)

In addition to `finetune_modules` and `base_run_dir`, you can specify any other config argument in the finetune config file, and it will override whatever was defined in the original config file. E.g., you'll probably want to set the number of epochs and learning rate schedule. In your case, as you're trying to finetune on a single basin, you'd also change `{train,validation,test}_basin_file`.

That said, clearly you can only change so much: If you change the model architecture (e.g., `hidden_size`), things will break because we won't be able to load the base model's weights into the finetune model.

**Comment by kratzert** [comment]
Not much too add besides that we already plan(ned) to add some tutorial(s) to explain stuff like fine-tuning, how to do PUB, how to continue training from checkpoint etc. But those things take time to write and we are quite busy with other stuff, so that might take a bit. 

I will close this issue but feel free to re-open if more questions arise.
